src/pokenav_unk_9.c

```diff
@@ -163,18 +186,541 @@ u32 sub_81CFA88(struct PokenavSub9 *structPtr)
+                ptr->unk10 = monId;
+                return LT_CONTINUE;
+            }
+        }
+        monId = 0;
+        boxId++;
+    }
+
+    ptr->unk14 = 1;
+    return LT_FINISH;
+}
+
+static void sub_81CFCEC(struct PokenavSub9 *structPtr, struct PokenavMonList *item)
+{
+    u32 left = 0;
+    u32 right = structPtr->unk1C->unk0;
+    u32 insertionIdx = left + (right - left) / 2;
+
+    while (right != insertionIdx)
+    {
+        if (item->unk6 > structPtr->unk1C->unk4[insertionIdx].unk6)
+            right = insertionIdx;
+        else
+            left = insertionIdx + 1;
+        insertionIdx = left + (right - left) / 2;
+    }
+    for (right = structPtr->unk1C->unk0; right > insertionIdx; right--)
+        structPtr->unk1C->unk4[right] = structPtr->unk1C->unk4[right - 1];
+    structPtr->unk1C->unk4[insertionIdx] = *item;
+    structPtr->unk1C->unk0++;
+}
+
+static bool32 sub_81CFD58(void)
+{
+    s32 i, j;
+
+    for (i = 0; i < PARTY_SIZE; i++)
+    {
+        struct Pokemon * mon = &gPlayerParty[i];
+        if (!GetMonData(mon, MON_DATA_SANITY_HAS_SPECIES))
+            continue;
+        if (GetMonData(mon, MON_DATA_SANITY_IS_EGG))
+            continue;
+        if (GetMonData(mon, MON_DATA_RIBBONS))
+            return TRUE;
+    }
+
+    for (i = 0; i < TOTAL_BOXES_COUNT; i++)
+    {
+        for (j = 0; j < IN_BOX_COUNT; j++)
+        {
+            if (!CheckBoxMonSanityAt(i, j))
+                continue;
+            if (GetBoxMonDataAt(i, j, MON_DATA_RIBBONS))
+                return TRUE;
+        }
+    }
+
+    return FALSE;
+}
+
+bool32 sub_81CFDD0(void)
+{
+    struct PokenavSub10 * ptr = AllocSubstruct(10, sizeof(struct PokenavSub10));
+    if (ptr == NULL)
+        return FALSE;
+    ptr->ltid = CreateLoopedTask(sub_81CFEB8, 1);
+    ptr->callback = sub_81CFE84;
+    ptr->unkC = FALSE;
+    return TRUE;
+}
+
+bool32 sub_81CFE08(void)
+{
+    struct PokenavSub10 * ptr = AllocSubstruct(10, sizeof(struct PokenavSub10));
+    if (ptr == NULL)
+        return FALSE;
+    ptr->ltid = CreateLoopedTask(sub_81CFEB8, 1);
+    ptr->callback = sub_81CFE84;
+    ptr->unkC = TRUE;
+    return TRUE;
+}
+
+void sub_81CFE40(s32 idx)
+{
+    struct PokenavSub10 * ptr = GetSubstructPtr(10);
+    ptr->ltid = CreateLoopedTask(gUnknown_086237B8[idx], 1);
+    ptr->callback = sub_81CFE84;
+}
+
+bool32 sub_81CFE70(void)
+{
+    struct PokenavSub10 * ptr = GetSubstructPtr(10);
+    return ptr->callback();
+}
+
+bool32 sub_81CFE84(void)
+{
+    struct PokenavSub10 * ptr = GetSubstructPtr(10);
+    return IsLoopedTaskActive(ptr->ltid);
+}
+
+void sub_81CFE98(void)
+{
+    struct PokenavSub10 * ptr = GetSubstructPtr(10);
+    sub_81C8234();
+    RemoveWindow(ptr->winid);
+    FreePokenavSubstruct(10);
+}
+
+static u32 sub_81CFEB8(s32 state)
+{
+    struct PokenavSub10 * unk = GetSubstructPtr(10);
+    switch (state)
+    {
+    case 0:
+        InitBgTemplates(gUnknown_086237B0, NELEMS(gUnknown_086237B0));
```

Comment:
Why NELEMS even exists, if we have ARRAY_COUNT?

Reply:
As discussed in the Discord, `NELEMS` is a GameFreak macro which appears in FireRed and Emerald-J asserts. It is identical to `ARRAY_COUNT`, and I aliased it as such in `global.h`.